### PR TITLE
Move col2im offset and stride to long to avoid overflows leading to corruption

### LINKED ISSUE #185663: [MPS] Col2Im.metal silent uint32 overflow in corrupts large outputs on MPS
### 🐛 Describe the bug

at::col2im on MPS produces incorrect results when K * L > 2³², where K = C·KH·KW and L = H_out·W_out are the channel and spatial extents of the input col tensor. The corrupted positions are those whose computed col_index would exceed 2³².

The problem happens at

```
  uint col_index = dim1_idx * col_channel_stride + dim2_idx * col_spatial_stride;
```

Opening a PR to move ulong for the relevant variables. This also mirrors what the op counterpart `at::im2col` already does.

Repro script:
```
import torch

KH, KW = 1, 65537
H_OUT, W_OUT = 1, 65537
H_IN, W_IN = 1, 131073           # so W_OUT = W_IN - KW + 1
K = KH * KW                      # 65537 (with C_in=1)
L = H_OUT * W_OUT                # 65537
# K * L = 65537 * 65537 = 2^32 + 131074  -> uint32 overflow inside col2im

torch.manual_seed(0)
cols = torch.randn(1, K, L)

out_cpu = torch.ops.aten.col2im(cols, [H_IN, W_IN], [KH, KW], [1, 1], [0, 0], [1, 1])

print("Running aten::col2im on MPS", flush=True)
out_mps = torch.ops.aten.col2im(
    cols.to("mps"), [H_IN, W_IN], [KH, KW], [1, 1], [0, 0], [1, 1]
).cpu()
torch.mps.synchronize()

print()
print(f"{'w':>10}  {'cpu':>14}  {'mps':>14}  {'abs_err':>14}")
for i in [0, 32768, 65535, 65536, 65537, 65538, 98304, 131072]:
    c = out_cpu.flatten()[i].item()
    m = out_mps.flatten()[i].item()
    print(f"{i:>10}  {c:>14.4f}  {m:>14.4f}  {abs(c - m):>14.4f}")

left_err = (out_cpu[..., :65536] - out_mps[..., :65536]).abs().max().item()
right_err = (out_cpu[..., 65536:] - out_mps[..., 65536:]).abs().max().item()
print(f"positions [0, 65535]      max abs err = {left_err:.4e}")
print(f"positions [65536, 131072] max abs err = {right_err:.4e}")
```

Printout

```
           w             cpu             mps         abs_err
           0         -1.1258         -1.1258          0.0000
       32768        214.7137        214.7132          0.0005
       65535       -256.6318       -256.6317          0.0000
       65536        -33.2780        -34.9529          1.6750
       65537       -300.6619       -302.2877          1.6257
       65538       -298.3085       -297.8763          0.4323
       98304       -224.1576       -222.6533          1.5043
      131072          1.0863          1.5554          0.4691

  positions [0, 65535]      max abs err = 1.0498e-02
  positions [65536, 131072] max abs err = 8.8287e+00
```

### Versions

Nightly, M2 Max, MacOS 26.4

cc @ezyang @gchanan @kadeng @msaroufim @kulinseth @malfet @DenisVieriu97 @aditvenk

### PR DESCRIPTION
**Pull request by jhavukainen**
Fixes https://github.com/pytorch/pytorch/issues/185663

Moves the offset / stride variables to use ulong to avoid silently overflowing uint32 for large inputs avoiding the overflow. This matches what the `at::im2col` implementation also does with promoting the relevant variables to 64-bits.

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/Col2Im.metal [hunk 1]
```diff
@@ -50,9 +50,9 @@ kernel void col2im_kernel(
   uint h_col_end = min((h_im / stride_h + 1), height_col);
 
   float accumulator = 0.0;
-  uint col_batch_offset = batch_idx * col_batch_stride;
-  uint col_channel_stride = col_inner_strides.x;
-  uint col_spatial_stride = col_inner_strides.y;
+  ulong col_batch_offset = batch_idx * col_batch_stride;
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/kernels/Col2Im.metal]: Does this hurt performance at all or should these indices be templated?

File: aten/src/ATen/native/mps/operations/Col2Im.mm
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/Col2Im.mm]: Feel like a lot of these other params could be problematic too with theoretical inputs now.

File: aten/src/ATen/native/mps/kernels/Col2Im.metal (hunk 1 above)
Comment (jhavukainen) [reply aten/src/ATen/native/mps/kernels/Col2Im.metal]: That's a valid question. Based on a similar stride computation change done here https://github.com/pytorch/pytorch/pull/182054#issuecomment-4356254356 where the couple of extra cycles spent in doing the address calculation with 64-bits get pretty well hidden my assumption is that it wont have a significant impact. And I didn't want to differ too much from the Im2Col implementation in this PR.

But this is also a good point to verify that and if there is an adverse effect we should templatize both Col2Im and Im2Col while we are at it. I'll run some numbers.

Comment (jhavukainen) [reply aten/src/ATen/native/mps/kernels/Col2Im.metal]: Ok with a naive benchmark focusing on the small shapes where the memory costs cant help hiding it do show 10-20% regression when moving to 64-bit stride computations. I think it's good enough reason to add the templating here since the extra cost is just more lines of code. I'll leave templatizing Im2Col to a separate PR.

Test snippet to run before and after below should anyone need it.

```
import time
import torch

# (label, N, C, KH, KW, H_out, W_out, padding)
CASES = [
    ("conv-bwd 3x3 64ch 56x56 N=32", 32, 64, 3, 3,  56,  56, 1),
    ("conv-bwd 7x7 64ch 224x224",     1, 64, 7, 7, 224, 224, 3),
    ("alu 21x21 4ch 48x48",           1,  4, 21, 21, 48, 48, 10),
]

for name, N, C, KH, KW, H, W, p in CASES:
    L = (H + 2 * p - KH + 1) * (W + 2 * p - KW + 1)
    cols = torch.randn(N, C * KH * KW, L, device="mps")
    args = ([H, W], [KH, KW], [1, 1], [p, p], [1, 1])
    for _ in range(5):
        torch.ops.aten.col2im(cols, *args)
    torch.mps.synchronize()
    t0 = time.perf_counter()
    iters = 30
    for _ in range(iters):
        torch.ops.aten.col2im(cols, *args)
    torch.mps.synchronize()
    print(f"{name:<32s}  {(time.perf_counter() - t0) / iters * 1e3:6.3f} ms/call")
```

File: aten/src/ATen/native/mps/operations/Col2Im.mm
Comment (jhavukainen) [reply aten/src/ATen/native/mps/operations/Col2Im.mm]: I think the stride params are a bit special here as the problem is overflowing the computation

`col_index = dim1_idx * col_channel_stride + dim2_idx * col_spatial_stride`

where we'll hit the uint32 limit with fairly mundane due to the multiplication. For the other params having values that end up crossing the limit would require significantly larger sizes, usually the 2^32 elements per dim limit starts to break a lot of the ops requiring maybe a more systematic review across the backend. But if we do have concrete cases that are being hit in col2im we should address it, the only one I'm currently encountering is with the stride computation.

File: aten/src/ATen/native/mps/kernels/Col2Im.metal (hunk 1 above)
Comment (jhavukainen) [reply aten/src/ATen/native/mps/kernels/Col2Im.metal]: Added a commit to templatize the kernel and dispatch suitably depending on if the stride computation will overflow uint32.

File: aten/src/ATen/native/mps/operations/Col2Im.mm [hunk 2]
```diff
@@ -72,9 +72,17 @@ static void col2im_out_mps_template(const Tensor& input,
   auto width_col = (output_width + 2 * pad_width - (dilation_width * (kernel_width - 1) + 1)) / stride_width + 1;
 
   auto stream = getCurrentMPSStream();
+  const int64_t max_col_offset = (batch_size > 0 ? (batch_size - 1) * input_batch_stride : 0) +
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Col2Im.mm]: I feel like there should already be a generic macro that returns tensor_size_in_storage

**Review by malfet — CHANGES_REQUESTED** [review]
Why just fix it to Col2Im, same bug equally affects Im2Col, isn't it?

Also, please add unit test (using char tensors for example)
Also, wouldn't this still be problematic if output.numel() exceed 2**32?

File: aten/src/ATen/native/mps/operations/Col2Im.mm (hunk 2 above)
Comment (jhavukainen) [reply aten/src/ATen/native/mps/operations/Col2Im.mm]: Yeah you're right there's the ` canUse32BitIndexMath` that should simplify these checks significantly. Added the change.

**Comment by jhavukainen** [comment]
@malfet: Im2Col already uses 64-bit types for strides and the relevant computations so it would not be affected by the original issue:
https://github.com/pytorch/pytorch/blob/5cd392bfe432d57e7beb9ab67037ddc0fcc01205/aten/src/ATen/native/mps/kernels/Im2Col.metal#L3-L12

The performance issue from always using the 64-bit instead of having it templatized would affect the Im2Col too, I was planning on opening a separate PR for those since its a different op in a different file. But I can also address it here if you prefer?

Good catch on the output.numel()>2^32, you're right it will hit a corruption as well. I've pushed a commit also widening the `dim1_idx` and `dim2_idx` to address the problem. Added regression tests to check both modes of the overflow issue.

File: aten/src/ATen/native/mps/operations/Col2Im.mm [hunk 3]
```diff
@@ -86,24 +89,36 @@ static void col2im_out_mps_template(const Tensor& input,
...
-          std::array<uint32_t, 2>{static_cast<uint32_t>(pad_height), static_cast<uint32_t>(pad_width)}, // pad_hw
-          std::array<uint32_t, 2>{static_cast<uint32_t>(stride_height),
-                                  static_cast<uint32_t>(stride_width)}, // stride_hw
-          std::array<uint32_t, 2>{static_cast<uint32_t>(dilation_height),
-                                  static_cast<uint32_t>(dilation_width)}, // dilation_hw
-          std::array<uint32_t, 2>{static_cast<uint32_t>(height_col), static_cast<uint32_t>(width_col)}, // col_hw
-          output_batch_stride,
-          std::array<uint32_t, 2>{static_cast<uint32_t>(input_channel_stride),
-                                  static_cast<uint32_t>(input_spatial_stride)}); // col_inner_strides
+      auto dispatch_args = [&](auto in_stride, auto out_stride, auto inner_strides) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Col2Im.mm:92]: Heh, didn't know it's a generic

**Review by malfet — APPROVED** [review]
